File: 3775-separate-squares-ii/separate-squares-ii.java
import java.util.*;

class Solution {
    // Segment Tree components
    private int[] count;
    private double[] length;
    private double[] X;

    private void update(int node, int start, int end, int l, int r, int val) {
        if (l > end || r < start) return;
        if (l <= start && end <= r) {
            count[node] += val;
        } else {
            int mid = start + (end - start) / 2;
            update(node * 2, start, mid, l, r, val);
            update(node * 2 + 1, mid + 1, end, l, r, val);
        }

        if (count[node] > 0) {
            length[node] = X[end + 1] - X[start];
        } else {
            if (start != end) {
                length[node] = length[node * 2] + length[node * 2 + 1];
            } else {
                length[node] = 0.0;
            }
        }
    }

    public double separateSquares(int[][] squares) {
        // 1. Coordinate Compression
        TreeSet<Double> xSet = new TreeSet<>();
        for (int[] sq : squares) {
            xSet.add((double) sq[0]);
            xSet.add((double) sq[0] + sq[2]);
        }
        
        X = new double[xSet.size()];
        int i = 0;
        for (Double val : xSet) X[i++] = val;
        
        Map<Double, Integer> xMap = new HashMap<>();
        for (i = 0; i < X.length; i++) xMap.put(X[i], i);

        int m = X.length;
        count = new int[4 * m];
        length = new double[4 * m];

        // 2. Events
        List<Event> events = new ArrayList<>();
        for (int[] sq : squares) {
            double y = sq[1];
            double l = sq[2];
            int idx1 = xMap.get((double)sq[0]);
            int idx2 = xMap.get((double)sq[0] + l);
            // Update interval [idx1, idx2 - 1]
            events.add(new Event(y, 1, idx1, idx2 - 1));
            events.add(new Event(y + l, -1, idx1, idx2 - 1));
        }
        
        Collections.sort(events);

        // 3. Sweep
        List<Strip> history = new ArrayList<>();
        double prev_y = events.get(0).y;
        
        for (Event ev : events) {
            double curr_y = ev.y;
            double dy = curr_y - prev_y;
            
            if (dy > 0) {
                history.add(new Strip(prev_y, dy, length[1]));
            }
            
            update(1, 0, m - 2, ev.x_l, ev.x_r, ev.type);
            prev_y = curr_y;
        }
        
        // 4. Calculate Total Area and find Split
        double totalArea = 0;
        for (Strip s : history) totalArea += s.dy * s.width;
        
        double target = totalArea / 2.0;
        double currentSum = 0;
        
        for (Strip s : history) {
            double stripArea = s.dy * s.width;
            if (currentSum + stripArea >= target) {
                double missing = target - currentSum;
                if (s.width == 0) return s.y_start;
                return s.y_start + (missing / s.width);
            }
            currentSum += stripArea;
        }
        
        return prev_y;
    }
    
    // Helper classes
    private static class Event implements Comparable<Event> {
        double y;
        int type;
        int x_l, x_r;
        public Event(double y, int type, int x_l, int x_r) {
            this.y = y; this.type = type; this.x_l = x_l; this.x_r = x_r;
        }
        public int compareTo(Event other) {
            return Double.compare(this.y, other.y);
        }
    }
    
    private static class Strip {
        double y_start, dy, width;
        public Strip(double y_start, double dy, double width) {
            this.y_start = y_start; this.dy = dy; this.width = width;
        }
    }
}

File: 3775-separate-squares-ii/separate-squares-ii.py
class Solution:
    def separateSquares(self, squares: List[List[int]]) -> float:
        # 1. Coordinate Compression for X
        xs = set()
        for x, y, l in squares:
            xs.add(x)
            xs.add(x + l)
        
        sorted_x = sorted(list(xs))
        x_map = {val: i for i, val in enumerate(sorted_x)}
        m = len(sorted_x)
        
        # Segment Tree Arrays
        # Range of elementary intervals is 0 to m-2
        tree_count = [0] * (4 * m)
        tree_len = [0.0] * (4 * m)
        
        def update(node, start, end, l, r, val):
            if l > end or r < start:
                return
            if l <= start and end <= r:
                tree_count[node] += val
            else:
                mid = (start + end) // 2
                update(node * 2, start, mid, l, r, val)
                update(node * 2 + 1, mid + 1, end, l, r, val)
            
            if tree_count[node] > 0:
                tree_len[node] = sorted_x[end + 1] - sorted_x[start]
            else:
                if start != end:
                    tree_len[node] = tree_len[node * 2] + tree_len[node * 2 + 1]
                else:
                    tree_len[node] = 0.0

        # 2. Events
        events = []
        for x, y, l in squares:
            idx1 = x_map[x]
            idx2 = x_map[x + l]
            # interval [idx1, idx2 - 1]
            events.append((y, 1, idx1, idx2 - 1))
            events.append((y + l, -1, idx1, idx2 - 1))
        
        events.sort(key=lambda x: x[0])
        
        # 3. Sweep
        history = [] # Stores (y_start, dy, width)
        prev_y = events[0][0]
        
        for y, type_, l_idx, r_idx in events:
            dy = y - prev_y
            if dy > 0:
                history.append((prev_y, dy, tree_len[1]))
            
            update(1, 0, m - 2, l_idx, r_idx, type_)
            prev_y = y
            
        # 4. Find Split
        total_area = sum(h[1] * h[2] for h in history)
        target = total_area / 2.0
        
        current_sum = 0.0
        for y_start, dy, width in history:
            strip_area = dy * width
            if current_sum + strip_area >= target:
                if width == 0:
                    return float(y_start)
                missing = target - current_sum
                return y_start + (missing / width)
            current_sum += strip_area
            
        return float(prev_y)